push notification support

diegogarciar opened the pull request:
Hello, 
I really like your plugin, thank you again for developing it. It was lacking push notification functionality which was crucial for me. So instead of just forking, I'm submitting this pull request to give back.

This pull request has 3 main changes

- Support for android level 16 just as Zendesk supports it #17 
- I mimic the behavior of iOS and Android for online agents #18 
- Added support for push notifications.

I tested the push notifications rigorously as I'll be the zendesk support agent and I cannot accept tickets immediately. So I'll rely heavily on users receiving push notifications. 

I also wanted added methods do connect and disconnect the web socket without deinitializing everything, I think those methods should be called when entering background and going back to foreground while the chat UI is loaded. This will help preserving battery life.

Lastly, I wanted to display a badge when there are new messages. My original approach was starting the `chatProviders`, but calling this method also 'connects' the web socket, so the user is considered 'connected' by zendesk. Zendesk doesn't send push notifications when the user is connected. To overcome this I added an optional parameter `chat.startChatProviders(connect: false)` (with default set to true so it doesn't change any behavior). This helped me initializing the observers and then connecting and disconnecting with a timer to check for new messages. I sent a support ticket to zendesk to check if there's a different solution as they claim it should be possible

https://develop.zendesk.com/hc/en-us/articles/360048210434-Adding-a-chat-message-counter-to-your-app-using-the-Chat-SDK

I documented all of it. 
The only lacking is creating a demo FirebaseMessagingService.kt  that users can declare in their AndroidManifest to display Android Push notifications. 

Please let me know your comments

KohlsAdrian commented:
Nice job! I am taking a look at it!

KohlsAdrian commented:
First thing that I need you to change is remove .lock file from the commit, maintaining .lock file prevents the plugin from getting newer version at fresh start, every time someone forks it, would have to delete it to force an upgrade

KohlsAdrian commented on lib/chat2/zendesk_chat2.dart:
```diff
@@ -51,7 +51,7 @@ class Zendesk2Chat {
     Map arguments = {
       'accountKey': accountKey,
       'appId': appId,
-      if (iosThemeColor != null) 'iosThemeColor': iosThemeColor.value,
```
you should not remove the conditional `iosThemeColor`. Even if it is obsolete, it may crash if the value is `null`, please revert this

diegogarciar replied:
Hello,
With the default value you are setting and null safety it can no longer be null.
The lint marked it as a unnecessary check for null

If you declared it as Color? it would be necessary
 ``` dart
Future<void> init(
    String accountKey,
    String appId, {
    @Deprecated('Prefer to use custom UI chat providers')
        Color? iosThemeColor = Colors.indigo,
  })
```

 do you still want me to revert the changes?

KohlsAdrian commented on android/src/main/kotlin/br/com/adriankohls/zendesk2/Zendesk2Chat.kt:
```diff
@@ -196,18 +204,9 @@ class Zendesk2Chat(private val activity: Activity?, private val channel: MethodC
     private fun accountProviderStart() {
         accountProviderObservationToken = ObservationScope()
         Chat.INSTANCE.providers()?.accountProvider()?.observeAccount(accountProviderObservationToken!!) {
-            when (it.status) {
```
You should revert this too.

these observer updated connection status and if it has online agents, doesn't mean that chat is connected that it will always have online agents, this is wrong

diegogarciar replied:
This is what I was referring in the issue, I mimicked the behavior from iOS. Or do you know if there's a variable to help determine if there are online agents even though no one has been assigned to the conversation?

https://github.com/KohlsAdrian/zendesk2/blob/3afddd04c76d5466075db9beea3048a6de57fcab/ios/Classes/SwiftZendesk2Chat.swift#L368

diegogarciar replied:
Well, I think I can simply use isOnline, I'll try that.

So about the inconsistency, do you want me to then revert this and change to this behavior in iOS?

KohlsAdrian commented on android/src/main/kotlin/br/com/adriankohls/zendesk2/Zendesk2Chat.kt:
```diff
@@ -179,7 +187,7 @@ class Zendesk2Chat(private val activity: Activity?, private val channel: MethodC
         chatProviderObservationToken = ObservationScope()
         Chat.INSTANCE.providers()?.chatProvider()?.observeChatState(chatProviderObservationToken!!) {
             this.agents = it.agents
-            this.hasAgents = it.agents.isNotEmpty()
```
You should revert this too.

these observer updated connection status and if it has online agents, doesn't mean that chat is connected that it will always have online agents, this is wrong

KohlsAdrian commented:
Added comments and waiting for changes.
Everything else LGTM!

KohlsAdrian commented:
And keeping the patterns, add the changelog.md as version `1.6.0` at the top and set what you have changed and set your as GitHub account as credits for helping

Eg:

## 1.6.0

* [What you have changed/made]

* Github user credits

## 1.5.0

diegogarciar commented:
Done, I also added the FirebaseMessagingService for reference but it is expected that users create their own. I also added that to the docs.

Lastly, here's the lint warning for the color check
<img width="979" alt="image" src="https://user-images.githubusercontent.com/17276155/118416678-20548580-b66e-11eb-9ae7-ddbf5f37b348.png">

KohlsAdrian commented:
Thank you, everything is perfect now, thank for all implementation and example at README.md

KohlsAdrian commented:
Releasing as 1.6.0 on pub.dev

diegogarciar commented:
You're welcome! Glad I could help!
